### PR TITLE
Add RepositoryFilesystem

### PR DESCRIPTION
**Pull request by mcmire**
When linting a project, we need to constantly read the files within either the module template or the project. We want this tool to run as fast as possible, and to do this, we can cache the metadata and content of each file that we read so that we don't have to do it again. This class assists with that.

Related to #5.

---

See here for context on how a RepositoryFilesystem will eventually be used:

- https://github.com/MetaMask/module-lint/blob/add-initial-code/src/rules/require-source-directory.ts#L10
- https://github.com/MetaMask/module-lint/blob/add-initial-code/src/establish-metamask-repository.ts#L89

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -9,6 +9,9 @@
     "noEmit": true,
     "noErrorTruncation": true,
     "noUncheckedIndexedAccess": true,
+    "paths": {
+      "@metamask/utils/node": ["./node_modules/@metamask/utils/dist/types/node"]
```
Comment (mcmire) [inline tsconfig.json:13]: The reason we're setting `paths` is explained here: https://github.com/MetaMask/module-lint/pull/20#discussion_r1373906853

File: src/misc-utils.test.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+              stack: expect.any(String),
+              cause: expect.objectContaining({
+                message: `EACCES: permission denied, stat '${filePath}'`,
+                code: 'EACCES',
+              }),
+            }),
+          );
+        } finally {
+          // Make sandbox root directory executable again
+          await fs.promises.chmod(sandbox.directoryPath, 0o700);
```
Comment (Gudahtt) [inline src/misc-utils.test.ts:58]: Nit: Is this step necessary? I thought using a sandbox removed the necessity to clean up; nothing will use this directory again after this test

File: src/repository-filesystem.test.ts [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+import { mock } from 'jest-mock-extended';
+import path from 'path';
+
+import { RepositoryFilesystem } from './repository-filesystem';
+import { withinSandbox } from '../tests/helpers';
+
+jest.mock('@metamask/utils/node', () => {
+  return {
+    // eslint-disable-next-line @typescript-eslint/naming-convention
+    __esModule: true,
```
Comment (Gudahtt) [inline src/repository-filesystem.test.ts:16]: Interesting, why was this mock needed?

File: src/repository-filesystem.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    this.#entryStats[entryPath] = stats;
+    return stats;
+  }
+
+  /**
+   * Builds the full path to a file or directory within the repository from a
+   * partial path.
+   *
+   * @param entryPath - The path to the file or directory within the context of
+   * the repository (so, minus its directory).
```
Comment (Gudahtt) [inline src/repository-filesystem.ts:66]: Nit: Calling the path within the project a  "path relative to the repository root" would be a bit clearer. Saying it's the path "minus its directory" isn't quite true, and "partial path" leaves ambiguous which part is missing.

```suggestion
   * @param filePath - The path to the file relative to the repository root.
   * @returns The contents of the file.
   */
  async readFile(filePath: string): Promise<string> {
    const cachedContent = this.#fileContents[filePath];
    const content =
      cachedContent ?? (await readFile(this.#getFullPath(filePath))).trim();
    this.#fileContents[filePath] = content;
    return content;
  }

  /**
   * Retrieves stats for the given file or directory.
   *
   * @param entryPath - The path to the file relative to the repository root.
   * @returns The `fs.Stats` object with information about the entry.
   */
  async getEntryStats(entryPath: string): Promise<fs.Stats | null> {
    const cachedStats = this.#entryStats[entryPath];
    const stats =
      cachedStats ?? (await getEntryStats(this.#getFullPath(entryPath)));
    this.#entryStats[entryPath] = stats;
    return stats;
  }

  /**
   * Builds the full path to a file or directory within the repository from a
   * path relative to the root directory of the repository.
   *
   * @param entryPath - The path to the file relative to the repository root.
```

**Review by Gudahtt — DISMISSED** [review]
LGTM!

**Review by Gudahtt — DISMISSED** [review]

File: src/misc-utils.test.ts (hunk 2 above)
Comment (mcmire) [reply src/misc-utils.test.ts:59]: If I remove this line, the `chmod` above seems to prohibit the sandbox from removing itself. I tried to remove the sandbox directory manually and got:

```
rm: /path/to/temp/directory/module-lint-tests--1699398082717/test-directory: Permission denied
rm: /path/to/temp/directory/module-lint-tests--1699398082717: Permission denied
```

Perhaps `withinSandbox` should set the `execute` permissions on the directory before deleting it.

Anyway, it's a bug that we should address in `utils`. I've added a reminder here: [16d512e](https://github.com/MetaMask/module-lint/pull/21/commits/16d512ea81e6afc5537c6d8de87407c4be8aab5d)

**Comment by mcmire** [comment]
^ Addressed merge conflicts.

File: src/misc-utils.test.ts (hunk 2 above)
Comment (Gudahtt) [reply src/misc-utils.test.ts:59]: Ah interesting, OK

**Review by Gudahtt — APPROVED** [review]
LGTM!

**Pull request merged**